02a-dbc.md

```diff
@@ -0,0 +1,153 @@
+checking: int,>=0                                     for value: Instance of <type 'float'>: 9.1104335791443   
+Variables bound in inner context:
+~~~
+
+For the value of 83, although the caller obeyed the contract, the function does not
+return an integer value. It fails the post-condition and an exception is raised.
+
+### Extending Contracts
+
+Sometimes, the simple builtin syntax for defining contracts is not sufficient. In this case, contracts can be extended by
+defining a function that implements a new contract. For example, number theory tells us that all perfect squares
+end in a digit of 1,4,5,6, or 9 or end in an even number of zero digits. We can define a new contract that checks
+this condition
+
+~~~ {.python}
+@new_contract
+def endsOk(x):
+    ends14569 = x%10 in (1,4,5,6,9)
+    ends00 = int(round((log(x,10)))) % 2 == 0
+    if ends14569 or ends00:
+        return True
+    raise ValueError("%s doesn't end in 1,4,5,6 or 9 or even number of zeros"%x)
+~~~
+
+We can then use this function, *endsOk*, in a contract specification
+
+~~~ {.python}
+@contract(x='int,endsOk,>=0',returns='int,>=0')
+def psqrt2(x):
+    return int(sqrt(x))
+~~~
+
+Let's see what happens when we try to use this *psqrt2* function on a number that ends in an odd number of zeros.
+
+~~~ {.output}
+>>> psqrt2(49)
+7
+>>> psqrt2(1000)
+Traceback (most recent call last):
+  File "../foo.py", line 24, in <module>
+    print "Perfect square root of 1000 = %d"%psqrt2(1000)
+  File "<decorator-gen-3>", line 2, in psqrt2
+  File "/Library/Python/2.7/site-packages/PyContracts/contracts/main.py", line 253, in contracts_checker
+    raise e
+contracts.interface.ContractNotRespected: Breach for argument 'x' to psqrt2().
+1000 doesn't end in 1,4,5,6 or 9 or even number of zeros
+checking: callable()       for value: Instance of <type 'int'>: 1000   
+checking: endsOk           for value: Instance of <type 'int'>: 1000   
+checking: int,endsOk,>=0   for value: Instance of <type 'int'>: 1000   
+Variables bound in inner context:
+~~~
+
+### Performance Considerations
+
+Depending on the situation, checking validity of a contract can be expensive. For example, suppose a function
+is designed to perform a binary search on a sorted list of numbers. A pre-condition for the operation is that
+the list it is given to search is sorted. If the list is large, checking that it is properly sorted is expensive.
+In other words, contracts can impact performance. For this reason, it is desirable for callers to have a way to
+disable contract checks to avoid always paying whatever performance costs they incur. This can be accomplished
+either by setting an environment variable, DISABLE_CONTRACTS or by a call to contracts.disable_all() **before**
+any @contracts are processed by the Python interpreter. This allows developers to keep the checks in place while
+they are developing code and then disable them once they are sure the code is working as expected.
+
```

Comment:
This would be a good spot to explain when is a good time to write these contracts. Before you write the function code? while you write the function code? after you write the function code?